accepted/2022/enable-512-vectors.md

```diff
@@ -0,0 +1,518 @@
+    do
+    {
+        uint bitPos = (uint)BitOperations.TrailingZeroCount(mask) / sizeof(char);
+        sepListBuilder.Append((int)(offset + bitPos));
+        mask = BitOperations.ResetLowestSetBit(mask);
+    } while (mask != 0);
+}
+```
+
+However, this is cumbersome for the developer for two reasons: the first, is that because the bitmask is built from the most significant bit of each `byte` element in a vector, we first have to mask off every other element in the mask with `0x5555` to link this properly back to our vector of `ushort`; the second, is when we perform a `TrailingZeroCount` (to get the first pos of the first true element in the `byte` vector) we have to _map it back_ to an offset in the `ushort` vector (done by dividing `sizeof(char)).
+
+The following example shows how `VectorMask128` allows to perform the same functionality without dealing with these lower level details:
+
+```C#
+Vector128<ushort> vector = Vector128.LoadUnsafe(ref source, offset);
+VectorMask128<ushort> v1Eq = Vector128.EqualsByVectorMask(vector, v1);
+VectorMask128<ushort> v2Eq = Vector128.EqualsByVectorMask(vector, v2);
+VectorMask128<ushort> v3Eq = Vector128.EqualsByVectorMask(vector, v3);
+VectorMask128<ushort> cmp = (v1Eq | v2Eq | v3Eq);
+
+if (cmp != VectorMask128<ushort>.Zero)
+{
+    // Skip every other bit
+    do
+    {
+        uint elmPos = cmp.FirstIndexOf(true);
+        sepListBuilder.Append((int)(offset + elmPos));
+        cmp = cmp.SetElementCond(elmPos, false);
+    } while (cmp != 0);
+}
+```
+
+`VectorMask128<ushort>` properly links conditional SIMD processing with `Vector128<ushort>`. The JIT best determines how to lower the `VectorMask` API based on available hardware of the system; however, because `VectorMask` encodes conditional length and element type, it allows the JIT to lower the code optimally.
+
+The following lists the new classes we propose;
+
+
+| Class  | Associated `Vector` API | 
+| ------ | ---------- |
+| `VectorMask64<T>`  | `Vector64<T>`  | 
+| `VectorMask128<T>` | `Vector128<T>` | 
+| `VectorMask256<T>` | `Vector256<T>` | 
+| `VectorMask<T>`    | `Vector<T>`    |  
+
+The following lists the methods for `VectorMask<T>` API that allow for combining `VectorMask<T> into complex conditions, e.g., boolean logic:
+
+| Method  |
+| ------ | 
+| `VectorMask<T> VectorMask<T>.And(VectorMask<T>, VectorMask<T>)`  | 
+| `VectorMask<T> VectorMask<T>.Or(VectorMask<T>, VectorMask<T>)`  | 
+| `VectorMask<T> VectorMask<T>.Not(VectorMask<T>, VectorMask<T>)`  | 
+| `VectorMask<T> VectorMask<T>.Xor(VectorMask<T>, VectorMask<T>)`  | 
+
+The following lists the methods that allow to use `VectorMask<T>` for some lower-level processing typically done in SIMD code
+
+(OPEN: What else?)
+
+| Method  |
+| ------ | 
+| `VectorMask<T> VectorMask<T>.FirstIndexOf(VectorMask<T> mask, bool val)`  | 
+| `VectorMask<T> VectorMask<T>.SetElemntCond(VectorMask<T> mask, ulong pos, bool val)` | 
```

Comment:
Do you envision `LeadingZeroCount` here returning the actual leading zero count? Because I had in mind that `FirstIndexOf` would return an index into the VectorMask by type, e.g.,

this
```
VectorMask<short> mask = ...;
int idx = VectorMask<short>.FirstIndexOf(mask, true);
```

vs

```
VectorMask<short> mask = ...;
int idx = VectorMask<short>.LeadingZeroCount(mask) / Unsafe.SizeOf(short);
```

Reply:
I think it would be the `LeadingZeroCount` based on type.

That is a `VectorMask128<byte>` would assume a mask of `16-bits` and so the `lzcnt` would be "0-15". A `VectorMask128<int>` on the other hand would assume a mask of `4-bits` and so the `lzcnt` would be `0-3`.

If we instead always returned a 32-bit based `lzcnt` for `VectorMask128` it would be easier to use with `ExtractMostSignificantBits` but less usable with the variable sized mask and with the abstract mask more generally.

Reply:
Ok, so we are in agreement over the behavior roughly, just instead of allowing to check first index of true/false condition, we are working with the leading zero count per type.

I think that's pretty reasonable.